### PR TITLE
Fetch mempool, vote and block data

### PR DESCRIPTION
**Pull request by ademuanthony**
This PR adds the feature for fetching, storing and retrieving mempool data with number of transaction, fee and size.

image

It also fetch and store block and vote data

image

### CODE REVIEW

File: postgres/models/block.go [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+	result, err := exec.ExecContext(ctx, sql, args...)
+	if err != nil {
+		return 0, errors.Wrap(err, "models: unable to update all in block slice")
+	}
+
+	rowsAff, err := result.RowsAffected()
+	if err != nil {
+		return 0, errors.Wrap(err, "models: unable to retrieve rows affected all in update all block")
+	}
+	return rowsAff, nil
```
Comment (Sirmorrison) [inline postgres/models/block.go]: add a newline.

File: postgres/models/block.go [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+}
+
+// Upsert attempts an insert using an executor, and does an update or ignore on conflict.
+// See boil.Columns documentation for how to properly use updateColumns and insertColumns.
+func (o *Block) Upsert(ctx context.Context, exec boil.ContextExecutor, updateOnConflict bool, conflictColumns []string, updateColumns, insertColumns boil.Columns) error {
```
Comment (Sirmorrison) [inline postgres/models/block.go]: this function is pretty hard to read... please add newlines where appropriate from line 469-498.

File: postgres/models/block.go [hunk 3]
```diff
@@ -0,0 +1,711 @@
...
+	var err error
+
+	if !cached {
+		insert, ret := insertColumns.InsertColumnSet(
+			blockAllColumns,
+			blockColumnsWithDefault,
+			blockColumnsWithoutDefault,
+			nzDefaults,
+		)
+		update := updateColumns.UpdateColumnSet(
```
Comment (Sirmorrison) [inline postgres/models/block.go]: newline.

File: postgres/models/block.go [hunk 4]
```diff
@@ -0,0 +1,711 @@
...
+}
+
+// Upsert attempts an insert using an executor, and does an update or ignore on conflict.
+// See boil.Columns documentation for how to properly use updateColumns and insertColumns.
+func (o *Block) Upsert(ctx context.Context, exec boil.ContextExecutor, updateOnConflict bool, conflictColumns []string, updateColumns, insertColumns boil.Columns) error {
+	if o == nil {
+		return errors.New("models: no block provided for upsert")
+	}
+
+	nzDefaults := queries.NonZeroDefaultSet(blockColumnsWithDefault, o)
```
Comment (Sirmorrison) [inline postgres/models/block.go]: please rename `nzDefaults`. `nonZeroDefaults` could work better.

File: postgres/models/block.go [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+	blockUpsertCacheMut.RUnlock()
+
+	var err error
+
+	if !cached {
+		insert, ret := insertColumns.InsertColumnSet(
+			blockAllColumns,
+			blockColumnsWithDefault,
+			blockColumnsWithoutDefault,
+			nzDefaults,
```
Comment (Sirmorrison) [inline postgres/models/block.go]: same as above

File: postgres/models/block.go [hunk 6]
```diff
@@ -0,0 +1,711 @@
...
+		if updateOnConflict && len(update) == 0 {
+			return errors.New("models: unable to upsert block, could not build update column list")
+		}
+
+		conflict := conflictColumns
+		if len(conflict) == 0 {
+			conflict = make([]string, len(blockPrimaryKeyColumns))
+			copy(conflict, blockPrimaryKeyColumns)
+		}
+		cache.query = buildUpsertQueryPostgres(dialect, "\"block\"", updateOnConflict, ret, update, conflict, insert)
```
Comment (Sirmorrison) [inline postgres/models/block.go]: Add newline

File: postgres/models/block.go [hunk 7]
```diff
@@ -0,0 +1,711 @@
...
+		if len(ret) != 0 {
+			cache.retMapping, err = queries.BindMapping(blockType, blockMapping, ret)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	value := reflect.Indirect(reflect.ValueOf(o))
+	vals := queries.ValuesFromMapping(value, cache.valueMapping)
```
Comment (Sirmorrison) [inline postgres/models/block.go]: `vals` should either be `values` or `mapValues`

File: postgres/models/block_test.go [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+
+	if query.Query == nil {
+		t.Error("expected a query, got nothing")
+	}
+}
+
+func testBlocksDelete(t *testing.T) {
+	t.Parallel()
+
+	seed := randomize.NewSeed()
```
Comment (Sirmorrison) [inline postgres/models/block_test.go]: Add newline

File: postgres/models/mempool.go [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+			mempoolColumnsWithDefault,
+			mempoolColumnsWithoutDefault,
+			nzDefaults,
+		)
+
+		cache.valueMapping, err = queries.BindMapping(mempoolType, mempoolMapping, wl)
+		if err != nil {
+			return err
+		}
+		cache.retMapping, err = queries.BindMapping(mempoolType, mempoolMapping, returnColumns)
```
Comment (Sirmorrison) [inline postgres/models/mempool.go]: Add newline

File: postgres/models/mempool.go [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+
+		if len(cache.retMapping) != 0 {
+			queryReturning = fmt.Sprintf(" RETURNING \"%s\"", strings.Join(returnColumns, "\",\""))
+		}
+
+		cache.query = fmt.Sprintf(cache.query, queryOutput, queryReturning)
+	}
+
+	value := reflect.Indirect(reflect.ValueOf(o))
+	vals := queries.ValuesFromMapping(value, cache.valueMapping)
```
Comment (Sirmorrison) [inline postgres/models/mempool.go]: same as above

File: postgres/models/mempool.go [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+
+		if len(wl) == 0 {
+			return 0, errors.New("models: unable to update mempool, could not build whitelist")
+		}
+
+		cache.query = fmt.Sprintf("UPDATE \"mempool\" SET %s WHERE %s",
+			strmangle.SetParamNames("\"", "\"", 1, wl),
+			strmangle.WhereClause("\"", "\"", len(wl)+1, mempoolPrimaryKeyColumns),
+		)
+		cache.valueMapping, err = queries.BindMapping(mempoolType, mempoolMapping, append(wl, mempoolPrimaryKeyColumns...))
```
Comment (Sirmorrison) [inline postgres/models/mempool.go]: Add newline

File: postgres/models/mempool.go [hunk 12]
```diff
@@ -0,0 +1,736 @@
...
+}
+
+// Upsert attempts an insert using an executor, and does an update or ignore on conflict.
+// See boil.Columns documentation for how to properly use updateColumns and insertColumns.
+func (o *Mempool) Upsert(ctx context.Context, exec boil.ContextExecutor, updateOnConflict bool, conflictColumns []string, updateColumns, insertColumns boil.Columns) error {
+	if o == nil {
+		return errors.New("models: no mempool provided for upsert")
+	}
+
+	nzDefaults := queries.NonZeroDefaultSet(mempoolColumnsWithDefault, o)
```
Comment (Sirmorrison) [inline postgres/models/mempool.go]: same as above... Rename.

File: postgres/models/mempool.go [hunk 13]
```diff
@@ -0,0 +1,736 @@
...
+
+	// Build cache key in-line uglily - mysql vs psql problems
+	buf := strmangle.GetBuffer()
```
Comment (Sirmorrison) [inline postgres/models/mempool.go]: Add newlines between line 500, 504, 509, 514, 518, 534, 548,  to make codes more readable.

File: postgres/models/vote.go [hunk 14]
```diff
@@ -0,0 +1,656 @@
...
+// FindVote retrieves a single record by ID with an executor.
+// If selectCols is empty Find will return all columns.
+func FindVote(ctx context.Context, exec boil.ContextExecutor, hash string, selectCols ...string) (*Vote, error) {
+	voteObj := &Vote{}
+
+	sel := "*"
+	if len(selectCols) > 0 {
+		sel = strings.Join(strmangle.IdentQuoteSlice(dialect.LQ, dialect.RQ, selectCols), ",")
+	}
+	query := fmt.Sprintf(
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: add newline

File: postgres/models/vote.go [hunk 15]
```diff
@@ -0,0 +1,656 @@
...
+// Insert a single record using an executor.
+// See boil.Columns.InsertColumnSet documentation to understand column list inference for inserts.
+func (o *Vote) Insert(ctx context.Context, exec boil.ContextExecutor, columns boil.Columns) error {
+	if o == nil {
+		return errors.New("models: no vote provided for insertion")
+	}
+
+	var err error
+
+	nzDefaults := queries.NonZeroDefaultSet(voteColumnsWithDefault, o)
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: same as above

File: postgres/models/vote.go [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+			voteColumnsWithDefault,
+			voteColumnsWithoutDefault,
+			nzDefaults,
+		)
+
+		cache.valueMapping, err = queries.BindMapping(voteType, voteMapping, wl)
+		if err != nil {
+			return err
+		}
+		cache.retMapping, err = queries.BindMapping(voteType, voteMapping, returnColumns)
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: newline

File: postgres/models/vote.go [hunk 17]
```diff
@@ -0,0 +1,656 @@
...
+
+		if len(cache.retMapping) != 0 {
+			queryReturning = fmt.Sprintf(" RETURNING \"%s\"", strings.Join(returnColumns, "\",\""))
+		}
+
+		cache.query = fmt.Sprintf(cache.query, queryOutput, queryReturning)
+	}
+
+	value := reflect.Indirect(reflect.ValueOf(o))
+	vals := queries.ValuesFromMapping(value, cache.valueMapping)
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: same as above. rename

File: postgres/models/vote.go [hunk 18]
```diff
@@ -0,0 +1,656 @@
...
+
+		if len(wl) == 0 {
+			return 0, errors.New("models: unable to update vote, could not build whitelist")
+		}
+
+		cache.query = fmt.Sprintf("UPDATE \"vote\" SET %s WHERE %s",
+			strmangle.SetParamNames("\"", "\"", 1, wl),
+			strmangle.WhereClause("\"", "\"", len(wl)+1, votePrimaryKeyColumns),
+		)
+		cache.valueMapping, err = queries.BindMapping(voteType, voteMapping, append(wl, votePrimaryKeyColumns...))
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: newline

File: postgres/models/vote.go [hunk 19]
```diff
@@ -0,0 +1,656 @@
...
+}
+
+// Upsert attempts an insert using an executor, and does an update or ignore on conflict.
+// See boil.Columns documentation for how to properly use updateColumns and insertColumns.
+func (o *Vote) Upsert(ctx context.Context, exec boil.ContextExecutor, updateOnConflict bool, conflictColumns []string, updateColumns, insertColumns boil.Columns) error {
+	if o == nil {
+		return errors.New("models: no vote provided for upsert")
+	}
+
+	nzDefaults := queries.NonZeroDefaultSet(voteColumnsWithDefault, o)
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: same ass above... Rename

File: postgres/models/vote.go [hunk 20]
```diff
@@ -0,0 +1,656 @@
...
+
+	// Build cache key in-line uglily - mysql vs psql problems
+	buf := strmangle.GetBuffer()
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: same as mentioned above.. line spaces where appropriate.

File: postgres/models/vote.go [hunk 21]
```diff
@@ -0,0 +1,656 @@
...
+
+	if len(cache.retMapping) != 0 {
+		err = exec.QueryRowContext(ctx, cache.query, vals...).Scan(returns...)
+		if err == sql.ErrNoRows {
+			err = nil // Postgres doesn't return anything when there's no update
+		}
+	} else {
+		_, err = exec.ExecContext(ctx, cache.query, vals...)
+	}
+	if err != nil {
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: newline

File: postgres/models/vote.go [hunk 22]
```diff
@@ -0,0 +1,656 @@
...
+		if updateOnConflict && len(update) == 0 {
+			return errors.New("models: unable to upsert vote, could not build update column list")
+		}
+
+		conflict := conflictColumns
+		if len(conflict) == 0 {
+			conflict = make([]string, len(votePrimaryKeyColumns))
+			copy(conflict, votePrimaryKeyColumns)
+		}
+		cache.query = buildUpsertQueryPostgres(dialect, "\"vote\"", updateOnConflict, ret, update, conflict, insert)
```
Comment (Sirmorrison) [inline postgres/models/vote.go]: newline

File: postgres/sqlboiler-sample.toml [hunk 23]
```diff
@@ -5,7 +5,9 @@
...
-        "vsp",
-        "vsp_tick",
-        "pow_data"]
\ No newline at end of file
+            "exchange_tick",
+            "vsp",
+            "vsp_tick",
+            "pow_data",
+            "mempool",
+            "block"]
```
Comment (Sirmorrison) [inline postgres/sqlboiler-sample.toml]: Add newline

File: web/server.go [hunk 24]
```diff
@@ -3,6 +3,7 @@ package web
 import (
 	"context"
 	"fmt"
+	"github.com/raedahgroup/dcrextdata/mempool"
```
Comment (Sirmorrison) [inline web/server.go]: re arrange imports

File: postgres/models/block.go [hunk 25]
```diff
@@ -0,0 +1,711 @@
...
+	mods = append(mods, qm.From("\"block\""))
+	return blockQuery{NewQuery(mods...)}
+}
+
+// FindBlock retrieves a single record by ID with an executor.
+// If selectCols is empty Find will return all columns.
+func FindBlock(ctx context.Context, exec boil.ContextExecutor, height int, selectCols ...string) (*Block, error) {
+	blockObj := &Block{}
+
+	sel := "*"
```
[the remark this answers is not in the document]
Comment (ademuanthony) [reply postgres/models/block.go]: this is an auto-generated code, it's not meant to be modified

**Pull request merged**